lib/cadet_web/controllers/auth_controller.ex

```diff
@@ -1,40 +1,59 @@
 defmodule CadetWeb.AuthController do
   use CadetWeb, :controller
+  use PhoenixSwagger
 
   import Ecto.Changeset
 
-  use PhoenixSwagger
-
   alias Cadet.Accounts
+  alias Cadet.Accounts.Form.Login
+  alias Cadet.Accounts.IVLE
   alias Cadet.Auth.Guardian
-  alias Cadet.Accounts.Form
-
-  def create(conn, %{"login" => attrs}) do
-    changeset = Form.Login.changeset(%Form.Login{}, attrs)
-
-    if changeset.valid? do
-      login = apply_changes(changeset)
 
-      case Accounts.sign_in(login.email, login.password) do
-        {:ok, user} ->
-          {:ok, access_token, _} =
-            Guardian.encode_and_sign(user, %{}, token_type: "access", ttl: {1, :hour})
+  @doc """
+  Receives a /login request with valid attributes (`Login form`).
 
-          {:ok, refresh_token, _} =
-            Guardian.encode_and_sign(user, %{}, token_type: "refresh", ttl: {52, :weeks})
+  If the user is already registered in our database, simply return `Tokens`. If
+  the user has not been registered before, register the user, then return the
+  `Tokens`.
+  """
+  def create(conn, %{"login" => attrs}) do
+    changeset = Login.changeset(%Login{}, attrs)
+
+    with valid = changeset.valid?,
+         {:changes, login} when valid <- {:changes, apply_changes(changeset)},
+         {:fetch, {:ok, nusnet_id}} <- {:fetch, IVLE.fetch_nusnet_id(login.ivle_token)},
+         {:signin, {:ok, user}} <- {:signin, Accounts.sign_in(nusnet_id, login.ivle_token)} do
+      {:ok, access_token, _} =
+        Guardian.encode_and_sign(user, %{}, token_type: "access", ttl: {1, :hour})
+
+      {:ok, refresh_token, _} =
+        Guardian.encode_and_sign(user, %{}, token_type: "refresh", ttl: {52, :weeks})
+
+      render(
+        conn,
+        "token.json",
+        access_token: access_token,
+        refresh_token: refresh_token
+      )
+    else
+      {:changes, _} ->
+        send_resp(conn, :bad_request, "Missing parameter")
 
-          render(conn, "token.json", access_token: access_token, refresh_token: refresh_token)
+      {:fetch, {:error, reason}} ->
+        # reason can be :bad_request or :internal_server_error
+        send_resp(conn, reason, "Unable to fetch NUSNET ID from IVLE.")
 
-        {:error, _reason} ->
-          send_resp(conn, :forbidden, "Wrong email and/or password")
-      end
-    else
-      send_resp(conn, :bad_request, "Missing parameters")
+      {:signin, {:error, reason}} ->
```

Comment:
This clause is missing test coverage hmm

Reply:
Done with eaa77fc.